src/include/miopen/tensor.hpp

```diff
@@ -173,6 +173,9 @@ struct TensorDescriptor : miopenTensorDescriptor
 
     std::string ToString() const;
 
+    void SetLayout(const std::string& l);
+    std::string GetLayout() const { return this->layout; }
```

Comment:
This is wrong as well. The layout should come from the order of the strides.

Reply:
Actually, the labels should be passed to `GetLayout` since the labels depend on the input type and not on the tensor layout. This can be implemented as:

```cpp
template <class Vector, class Op>
inline std::vector<int64_t> sort_permutation(const Vector& data, Op op)
{
    std::vector<std::int64_t> result(data.size());
    std::iota(result.begin(), result.end(), 0);
    std::sort(result.begin(), result.end(), [&](auto x, auto y) { return op(data[x], data[y]); });
    return result;
}

std::string TensorDescriptor::GetLayout(std::string labels) const
{
    auto result = labels;
    auto p = sort_permutation(strides, std::greater<>{});
    std::tranform(p.begin(), p.end(), result.begin(), [&](auto i) {
        return labels[i];
    });
    return result;
}
```

So then you can call it with `GetLayout("NCHW")` and if the layout is CNHW it will CNHW.

Reply:
Nice, it gives client the flexibility to specify the layout character strings too. I like this implementation.